### PR TITLE
Use Coroutines for TickTasks

### PR DESCRIPTION
**Pull request by My-Name-Is-Jeff**
Could completely remove ticktask maybe?

### CODE REVIEW

File: src/main/java/gg/skytils/skytilsmod/mixins/transformers/MixinMinecraft.java
Comment (Sychic) [inline src/main/java/gg/skytils/skytilsmod/mixins/transformers/MixinMinecraft.java:1]: Personally not a fan of mixing in and adding it

File: src/main/kotlin/gg/skytils/skytilsmod/utils/Utils.kt [hunk 1]
```diff
@@ -63,6 +63,7 @@ import java.io.File
 import java.io.IOException
 import java.util.*
 import kotlin.math.floor
+import kotlin.time.Duration.Companion.milliseconds
```
Comment (Sychic) [inline src/main/kotlin/gg/skytils/skytilsmod/utils/Utils.kt:66]: Unnecessary import?

File: src/main/kotlin/gg/skytils/skytilsmod/utils/tasks/TickTask.kt [hunk 2]
```diff
@@ -16,27 +16,50 @@
...
+import kotlinx.coroutines.launch
+import kotlin.coroutines.CoroutineContext
 
-class TickTask<T>(val ticks: Int = 0, val repeats: Boolean = false, register: Boolean = true, val task: () -> T) {
-    var remainingTicks = ticks
+class TickTask<T>(
+    val ticks: Int = 0,
+    val repeats: Boolean = false,
+    register: Boolean = true,
+    val task: (CoroutineContext) -> T
```
Comment (Sychic) [inline src/main/kotlin/gg/skytils/skytilsmod/utils/tasks/TickTask.kt:31]: Could add a secondary constructor that takes a lambda with no args

File: src/main/kotlin/gg/skytils/skytilsmod/utils/tasks/TickTask.kt [hunk 3]
```diff
@@ -16,27 +16,50 @@
...
+    fun unregister() {
+        job?.cancel() ?: error("Task is not registered!")
+        job = null
+    }
 
-    internal fun complete() = try {
-        callback(task())
+    private fun complete(ctx: CoroutineContext) = try {
+        callback(task(ctx))
     } catch (t: Throwable) {
```
Comment (Sychic) [inline src/main/kotlin/gg/skytils/skytilsmod/utils/tasks/TickTask.kt:63]: Probably want to catch `CoroutineCancellationException` separately

**Review by Sychic — CHANGES_REQUESTED** [review]
The behavior changes from running every tick to running every game loop

**Comment by My-Name-Is-Jeff** [comment]
> The behavior changes from running every tick to running every game loop

How does coroutine work? I think they accomplish the same thing unless Minecraft is running behind on ticks

**Comment by Sychic** [comment]
> > The behavior changes from running every tick to running every game loop
> 
> How does coroutine work? I think they accomplish the same thing unless Minecraft is running behind on ticks

Well since the coroutine scope has a MC executor as a dispatcher, it's going to queue the tasks there. AFAICT, MC's `addScheduledTask` executes every game loop as opposed to every game tick.

**Comment by My-Name-Is-Jeff** [comment]
> > > The behavior changes from running every tick to running every game loop
> > 
> > 
> > How does coroutine work? I think they accomplish the same thing unless Minecraft is running behind on ticks
> 
> Well since the coroutine scope has a MC executor as a dispatcher, it's going to queue the tasks there. AFAICT, MC's `addScheduledTask` executes every game loop as opposed to every game tick.

Delay behavior should be the same if each tick takes 50ms right? Should be fine because ticks also happen in game loop

**Comment by Sychic** [comment]
> > > > The behavior changes from running every tick to running every game loop
> > > 
> > > 
> > > How does coroutine work? I think they accomplish the same thing unless Minecraft is running behind on ticks
> > 
> > 
> > Well since the coroutine scope has a MC executor as a dispatcher, it's going to queue the tasks there. AFAICT, MC's `addScheduledTask` executes every game loop as opposed to every game tick.
> 
> Delay behavior should be the same if each tick takes 50ms right? Should be fine because ticks also happen in game loop

I think game loops aren't aligned to ticks since game loops are also where the game renders each frame. The delay behavior makes sense but it wouldn't be synced to a tick to begin with.

**Comment by My-Name-Is-Jeff** [comment]
Superseded by 7a741404e010ba19f6b52477341b5720e31c04a7

**Pull request closed without merging**